### PR TITLE
Avoid fetching OpenSSL digests and ciphers

### PR DESCRIPTION
**Pull request by paulidale**
The old cipher returning calls like `EVP_aes_128_gcm()` perform late binding which means they fetch on initialisation.  Fetching in OpenSSL 3.0 is a relatively expensive operation.  Instead of fetching every time a cipher is required, it is faster to pre-fetch and reuse the same `EVP_CIPHER` object.

Likewise, HMAC is better prefetched but it has the additional complexity of fetching a digest internally.  Instead of just prefetching the `EVP_MAC` object, it is better to create an `EVP_MAC_CTX` object and call `EVP_MAC_CTX_dup()` as required.

According to the profiling I've done, this represents a circa 4% boost in HPS.

### CODE REVIEW

File: src/platform/crypt_openssl.c
Comment (nibanks) [inline src/platform/crypt_openssl.c:58]:
```suggestion
    char *cipher_name,
    EVP_CIPHER **cipher
```

File: src/platform/crypt_openssl.c
Comment (nibanks) [inline src/platform/crypt_openssl.c:151]:
```suggestion
    if (!CxPlatLoadCipher("AES-128-GCM", &CXPLAT_AES_128_GCM_ALG_HANDLE) ||
        !CxPlatLoadCipher("AES-256-GCM", &CXPLAT_AES_256_GCM_ALG_HANDLE) ||
        !CxPlatLoadCipher("AES-128-ECB", &CXPLAT_AES_128_ECB_ALG_HANDLE) ||
        !CxPlatLoadCipher("AES-256-ECB", &CXPLAT_AES_256_ECB_ALG_HANDLE) ||
        !CxPlatLoadCipher("ChaCha20-Poly1305", &CXPLAT_CHACHA20_POLY1305_ALG_HANDLE)) {
        goto Error;
    }
```

File: src/platform/crypt_openssl.c
Comment (nibanks) [inline src/platform/crypt_openssl.c:167]:
```suggestion
    if (!CxPlatLoadHMACCTX(mac, "sha256", &CXPLAT_HMAC_SHA256_CTX_HANDLE) ||
        !CxPlatLoadHMACCTX(mac, "sha384", &CXPLAT_HMAC_SHA384_CTX_HANDLE) ||
        !CxPlatLoadHMACCTX(mac, "sha512", &CXPLAT_HMAC_SHA512_CTX_HANDLE)) {
        goto Error;
    }
```

File: src/platform/crypt_openssl.c
Comment (nibanks) [inline src/platform/crypt_openssl.c:608]:
```suggestion
    case CXPLAT_HASH_SHA384:
```

File: src/platform/crypt_openssl.c
Comment (nibanks) [inline src/platform/crypt_openssl.c:611]:
```suggestion
    case CXPLAT_HASH_SHA512:
```

File: src/platform/crypt_openssl.c
Comment (nibanks) [inline src/platform/crypt_openssl.c:621]: For allocation failures, please log the following type of event:
```
            QuicTraceEvent(
                AllocFailure,
                "Allocation of '%s' failed. (%llu bytes)",
                "EVP_MAC_CTX_dup",
                0); // If you don't know the size, just put zero
```
I'd appreciate it if you could add these to all initial points of allocation failure.

File: src/platform/crypt_openssl.c
Comment (nibanks) [inline src/platform/crypt_openssl.c:37]: codecheck build is complaining:
```
/home/vsts/work/1/s/src/platform/crypt_openssl.c:30:1: error: #includes are not sorted properly [llvm-include-order,-warnings-as-errors]
#include "openssl/kdf.h"
^        ~~~~~~~~~~~~~~~
         "openssl/evp.h"
```
So, could you please put these in alphabetical order?

File: src/platform/crypt_openssl.c
Comment (nibanks) [inline src/platform/crypt_openssl.c:55]: Not sure why, but macOS static linking build is complaining about a bunch of duplicate symbols:
```
/Applications/Xcode_12.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -Og -gdwarf -arch x86_64 -isysroot /Applications/Xcode_12.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk -mmacosx-version-min=10.15 -Wl,-search_paths_first -Wl,-headerpad_max_install_names CMakeFiles/quicipserver.dir/quicipserver.cpp.o -o ../../../../../../../artifacts/bin/macos/x64_Debug_openssl_Static/quicipserver  ../../../../../../../artifacts/bin/macos/x64_Debug_openssl_Static/libmsquic.a -framework CoreFoundation -framework Security 
duplicate symbol '_ENGINE_new' in:
    ../../../../../../artifacts/bin/macos/x64_Debug_openssl_Static/libmsquic.a(eng_lib.o)
    ../../../_deps/opensslquic-build/openssl/lib/libcrypto.a(eng_lib.o)
duplicate symbol '_engine_cleanup_add_first' in:
    ../../../../../../artifacts/bin/macos/x64_Debug_openssl_Static/libmsquic.a(eng_lib.o)
    ../../../_deps/opensslquic-build/openssl/lib/libcrypto.a(eng_lib.o)
duplicate symbol '_engine_cleanup_add_last' in:
    ../../../../../../artifacts/bin/macos/x64_Debug_openssl_Static/libmsquic.a(eng_lib.o)
    ../../../_deps/opensslquic-build/openssl/lib/libcrypto.a(eng_lib.o)
duplicate symbol '_engine_cleanup_int' in:
    ../../../../../../artifacts/bin/macos/x64_Debug_openssl_Static/libmsquic.a(eng_lib.o)
    ../../../_deps/opensslquic-build/openssl/lib/libcrypto.a(eng_lib.o)
duplicate symbol '_engine_lock_init' in:
    ../../../../../../artifacts/bin/macos/x64_Debug_openssl_Static/libmsquic.a(eng_lib.o)
    ../../../_deps/opensslquic-build/openssl/lib/libcrypto.a(eng_lib.o)
duplicate symbol '_ENGINE_set_cmd_defns' in:
    ../../../../../../artifacts/bin/macos/x64_Debug_openssl_Static/libmsquic.a(eng_lib.o)
    ../../../_deps/opensslquic-build/openssl/lib/libcrypto.a(eng_lib.o)
...
```

File: src/platform/crypt_openssl.c
Comment (nibanks) [inline src/platform/crypt_openssl.c:181]: Seems the chacha20 unit tests are failing (cross-platform). Here's one example:
```
##[error][03/28/2022 11:26:42] CryptTest.WellKnownChaChaPolyv1 failed:
Note: Google Test filter = CryptTest.WellKnownChaChaPolyv1
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from CryptTest
[ RUN      ] CryptTest.WellKnownChaChaPolyv1
/home/vsts/work/1/s/src/platform/unittest/CryptTest.cpp:327: Failure
Expected equality of these values:
  0
  memcmp(HpMask, ExpectedHpMask.Data, ExpectedHpMask.Length)
    Which is: -1
[  FAILED  ] CryptTest.WellKnownChaChaPolyv1 (0 ms)
```

**Review by nibanks — CHANGES_REQUESTED** [review]
Thanks a lot for making these changes. I few things need a bit of clean up before we can merge this. Also, please verify you've signed the Microsoft CLA. Thanks!

**Comment by nibanks** [comment]
You also need to run `./scripts/update-sidecar.ps1` but only after all event code changes are made.

File: src/platform/crypt_openssl.c
Comment (thhous-msft) [reply src/platform/crypt_openssl.c:55]: This is an error on our build scripts. In src/tools/pcp/CMakeLists.txt, chane the target_link_libraries line to 

```
if (BUILD_SHARED_LIBS)
target_link_libraries(quicpcp core platform)
endif()
```

File: src/platform/crypt_openssl.c
Comment (paulidale) [reply src/platform/crypt_openssl.c:181]: No, they are all static structure in 1.1.1 and before.

File: src/platform/crypt_openssl.c
Comment (paulidale) [reply src/platform/crypt_openssl.c:37]: Sure.

File: src/platform/crypt_openssl.c
Comment (paulidale) [reply src/platform/crypt_openssl.c:486]: This is wrong, we don't need to NULL check anything here, a NULL for any of these will have failed during initialisation.

File: src/platform/crypt_openssl.c
Comment (paulidale) [reply src/platform/crypt_openssl.c:116]: Just keeping the old (uncalled) code around, I'll remove.

File: src/platform/crypt_openssl.c
Comment (paulidale) [reply src/platform/crypt_openssl.c:140]: FWIW

File: src/platform/crypt_openssl.c
Comment (paulidale) [reply src/platform/crypt_openssl.c:621]: Sure.

File: src/platform/crypt_openssl.c
Comment (paulidale) [reply src/platform/crypt_openssl.c:175]: This is weird, it's not clear why the changes I made in either use of this cipher would impact this test.  Not using a fetched cipher results in the same error - this diff:
```diff
diff --git a/src/platform/crypt_openssl.c b/src/platform/crypt_openssl.c
index d7d25e6c..c53d826d 100644
--- a/src/platform/crypt_openssl.c
+++ b/src/platform/crypt_openssl.c
@@ -241,7 +241,7 @@ CxPlatKeyCreate(
         Aead = CXPLAT_AES_256_GCM_ALG_HANDLE;
         break;
     case CXPLAT_AEAD_CHACHA20_POLY1305:
-        Aead = CXPLAT_CHACHA20_POLY1305_ALG_HANDLE;
+        Aead = (EVP_CIPHER *)EVP_chacha20_poly1305(); /*CXPLAT_CHACHA20_POLY1305_ALG_HANDLE;*/
         break;
     default:
         Status = QUIC_STATUS_NOT_SUPPORTED;
@@ -515,7 +515,7 @@ CxPlatHpKeyCreate(
         Aead = CXPLAT_AES_256_ECB_ALG_HANDLE;
         break;
     case CXPLAT_AEAD_CHACHA20_POLY1305:
-        Aead = CXPLAT_CHACHA20_POLY1305_ALG_HANDLE;
+        Aead = (EVP_CIPHER *)EVP_chacha20_poly1305(); /*CXPLAT_CHACHA20_POLY1305_ALG_HANDLE;*/;
         break;
     default:
         Status = QUIC_STATUS_NOT_SUPPORTED;
```

I switch to using `EVP_CIPHER_CTX_set_params()` instead of `EVP_CIPHER_CTX_ctrl()` but that doesn't make a difference either, but it demonstrates that it isn't the shim code between the latter and the former that's broken.

**Comment by paulidale** [comment]
> You also need to run `./scripts/update-sidecar.ps1` but only after all event code changes are made.

That exploded nicely :(

```
dotnet: /home/pauli/work/msquic/msquic/scripts/update-sidecar.ps1:45
Line |
  45 |  dotnet publish ../submodules/clog/src/clog -o ${ClogDir} -f net6.0
     |  ~~~~~~
     | The term 'dotnet' is not recognized as a name of a cmdlet,
     | function, script file, or executable program. Check the
     | spelling of the name, or if a path was included, verify that
     | the path is correct and try again.

/home/pauli/work/msquic/msquic/build/clog/clog: The term '/home/pauli/work/msquic/msquic/build/clog/clog' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
/home/pauli/work/msquic/msquic/build/clog/clog: The term '/home/pauli/work/msquic/msquic/build/clog/clog' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
/home/pauli/work/msquic/msquic/build/clog/clog: The term '/home/pauli/work/msquic/msquic/build/clog/clog' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
/home/pauli/work/msquic/msquic/build/clog/clog: The term '/home/pauli/work/msquic/msquic/build/clog/clog' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
/home/pauli/work/msquic/msquic/build/clog/clog: The term '/home/pauli/work/msquic/msquic/build/clog/clog' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
```

**Comment by paulidale** [comment]
> Also, please verify you've signed the Microsoft CLA.

The CLA is underway, I'll need an OMC vote before I can confirm that my employer is agreeable.  I've added this to the next meeting's agenda.

**Comment by thhous-msft** [comment]
You need .net 6 installed to run update-sidecar.ps1. https://docs.microsoft.com/en-us/dotnet/core/install/linux

File: src/platform/crypt_openssl.c
Comment (nibanks) [reply src/platform/crypt_openssl.c:175]: Could it be your removal of `OPENSSL_init_ssl(OPENSSL_INIT_LOAD_CONFIG, NULL)`? The tests are still failing with your latest changes.

**Comment by paulidale** [comment]
I've run the update script and pushed.
Still trying to figure out the ChaCha20-Poly1305 issue.

File: src/platform/crypt_openssl.c
Comment (anrossi) [inline src/platform/crypt_openssl.c:46]: Are these thread-safe?

File: src/platform/crypt_openssl.c
Comment (anrossi) [inline src/platform/crypt_openssl.c:94]: Are you familiar with SAL annotations?  If so, would you add the appropriate annotations to your new functions?

File: src/platform/crypt_openssl.c
Comment (paulidale) [reply src/platform/crypt_openssl.c:46]: Yes, they are readonly after they are created which is at initialisation time.

File: src/platform/crypt_openssl.c
Comment (paulidale) [reply src/platform/crypt_openssl.c:94]: No, I'm not familiar with SAL annotations...

File: src/platform/crypt_openssl.c
Comment (nibanks) [reply src/platform/crypt_openssl.c:175]: @paulidale not sure if you saw my previous comment, but I am guessing your removal of `OPENSSL_init_ssl` is what caused the failure for chacha20 tests.

Comment (paulidale) [reply src/platform/crypt_openssl.c:175]: No, I missed it.  The configuration should load automatically, but it's not problem to put the init call back.

File: src/platform/crypt_openssl.c
Comment (paulidale) [reply src/platform/crypt_openssl.c]: I've made a stab at adding the annotations.

File: src/platform/crypt_openssl.c
Comment (nibanks) [inline src/platform/crypt_openssl.c:157]: Our coding style requires `{` and `}` around all code blocks.
```suggestion
        !CxPlatLoadCipher("ChaCha20-Poly1305", &CXPLAT_CHACHA20_POLY1305_ALG_HANDLE)) {
        goto Error;
    }
```
Please update the other code too.

File: src/platform/crypt_openssl.c
Comment (nibanks) [reply src/platform/crypt_openssl.c:185]: Well it seems that the tests are still failing even with the latest changes. Really weird!

Comment (paulidale) [reply src/platform/crypt_openssl.c:188]: Not too surprising: the init call and config file load are automatic in 3.0 (& 1.1.1).

What is surprising is that a crypto algorithm is failing.  The reason eludes me currently.

File: src/platform/crypt_openssl.c
Comment (paulidale) [reply src/platform/crypt_openssl.c]: Done hopefully.

File: src/platform/crypt_openssl.c
Comment (nibanks) [inline src/platform/crypt_openssl.c:534]: @paulidale this is the diff that is causing the test failures. For Header Protection, we use just `chacha20`, not including the `Poly1305`.

Comment (paulidale) [reply src/platform/crypt_openssl.c:534]: Thanks, I'm must have been blind to not notice this.

**Comment by paulidale** [comment]
The TLS changes give me better than 2x improvement to handshakes/second.  There is another one coming that will (hopefully) be better.

**Comment by nibanks** [comment]
The perf results are in (below), and I don't see a significant difference in HPS. Is this PR primarily for 3.0 perf?

### Windows OpenSSL (1.1.1)
```
Run 1: 6073 HPS
Run 2: 6025 HPS
Run 3: 6183 HPS
Run 4: 6041 HPS
Run 5: 6192 HPS
Median: 6073 HPS (+.11%)
Remote: 6066.6 HPS
```
### Linux OpenSSL (1.1.1)
```
Run 1: 12672 HPS
Run 2: 12834 HPS
Run 3: 12497 HPS
Run 4: 13086 HPS
Run 5: 12887 HPS
Median: 12834 HPS (+1.23%)
Remote: 12678.4 HPS
```

**Comment by paulidale** [comment]
I missed the OMC meeting yesterday and the item slipped a week.

Yes, this is for 3.x performance.  I doubt it will make a noticeable change to 1.1.1.

The other change is avoid calls to `SSL_CTX_new()` which is a pretty heavy operation (more so in 3.x than in 1.1.1).  A new `SSL_CTX` object is currently being created for every connection whereas the usual approach is to just create one, configure it and create the `SSL` objects from that.  It's not quite that straightforward with SNI and ALPN where you might need several pre-created `SSL_CTX` objects swapping in the correct one using `SSL_set_SSL_CTX()`.

I'm not sure how to fit this into the library/performance test framework.  It might be better done as a separate PR: it will be a set of quite distinct changes to what is here and likely involve more widespread modifications in the OpenSSL platform code.

**Comment by paulidale** [comment]
An email discussion is unlikely to be faster.  That's why we have regular calls.  This Wednesday hopefully.

**Comment by nibanks** [comment]
> An email discussion is unlikely to be faster. That's why we have regular calls. This Wednesday hopefully.

Any updates on getting approval @paulidale?

**Comment by paulidale** [comment]
Not yet :(

**Comment by nibanks** [comment]
@paulidale any updates? Is there some particular hold up on getting approval to sign?

**Comment by paulidale** [comment]
The committee wants some input from our lawyers and this is but no means the highest priority.  It will happen, the question at the moment is when.

**Comment by paulidale** [comment]
Sadly, nope :cry:

**Comment by nibanks** [comment]
It's been another month @paulidale. We'd really appreciate if you could poke the folks you're blocked on to get resolution here. Thanks.

**Comment by paulidale** [comment]
Poking won't help.  The person who deals with the legals has more important concerns which aren't going away anytime soon.  We are working on getting an alternate.

I do apologise for the delay.

**Comment by nibanks** [comment]
It's been another month @paulidale. Do you have any idea how long until you can get approval? Would you please poke the necessary people? Thanks!

**Comment by paulidale** [comment]
Sorry, I've been on vacation.  I'll prod this week.

**Comment by paulidale** [comment]
Our admin person is looking into this now.  The CLA link up top is broken.

**Comment by nibanks** [comment]
Thanks for the update @paulidale. I think if you push a new commit, the link would probably refresh. BTW, you have a merge conflict with the latest main which needs to be fixed anyways.

**Comment by paulidale** [comment]
I've finally got permission to sign the CLA on behalf of the project.
The links above are still broken and it's not clear how to sign from the [CLA page](https://cla.opensource.microsoft.com/).

**Comment by nibanks** [comment]
Weird. Perhaps try to close and reopen the PR? Otherwise might have to recreate the PR. Or just create a different temp PR to get a new link.

**Comment by paulidale** [comment]
@microsoft-github-policy-service agree company="OpenSSL Software Services"

**Comment by paulidale** [comment]
Got it.  The entire process has changed and the CLA page hasn't.

**Comment by nibanks** [comment]
> Got it. The entire process has changed and the CLA page hasn't.

Great. Would you mind merging or rebasing on `main` to fix some CI issues, please? Then hopefully we can get this merged ASAP today. Thanks again!

**Comment by nibanks** [comment]
Know what, never mind. All the issues are fixed issues. I'm going to merge this. Thanks!

**Comment by paulidale** [comment]
Thanks.  I'm on holiday for a few days and wouldn't have got to it until next week.

BTW: there are other performance fixes in OpenSSL's master branch that would be very worthwhile here.

**Pull request merged**
